src/main/scala/com/microsoft/ml/spark/vw/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,421 @@
+
+  private def createLabelSetter(df: DataFrame) = {
+    val labelColIdx = df.schema.fieldIndex(getLabelCol)
+
+    if (get(weightCol).isDefined) {
+      val weightColIdx = df.schema.fieldIndex(getWeightCol)
+      (row: Row, ex: VowpalWabbitExample) =>
+        ex.setLabel(row.getDouble(weightColIdx).toFloat, row.getDouble(labelColIdx).toFloat)
+    }
+    else
+      (row: Row, ex: VowpalWabbitExample) => ex.setLabel(row.getDouble(labelColIdx).toFloat)
+  }
+
+  /**
+    * Generate namespace info (hash, feature group, field index) for supplied columns.
+    * @param schema data frame schema to lookup column indices.
+    * @return
+    */
+  private def generateNamespaceInfos(schema: StructType): Array[NamespaceInfo] =
+    (Seq(getFeaturesCol) ++ getAdditionalFeatures)
+      .map(col => new NamespaceInfo(VowpalWabbitMurmur.hash(col, getHashSeed), col.charAt(0), schema.fieldIndex(col)))
+      .toArray
+
+  /**
+    * If multiple passes are requested and cache file is not enabled, cache the data in-memory.
+    * @param inputRows input data.
+    * @return iterator over the potentially cached data.
+    */
+  private def createDataIterator(inputRows: Iterator[Row], numPassesInSpark: Int) = {
+    if (numPassesInSpark == 1)
+    // avoid caching inputRows
+      () => inputRows
+    else {
+      // cache all row references to be able to quickly get the iterator again
+      // TODO: switch dataframe.cache()
+      val inputRowsArr = inputRows.toArray
+      () => inputRowsArr.iterator
+    }
+  }
+
+  private def buildCommandLineArguments(vwArgs: String, contextArgs: => String = "") = {
+    val args = new StringBuilder
+    args.append(vwArgs)
+      .append(" ").append(contextArgs)
+
+    // have to pass to get multi-pass to work
+    val no_stdin = "--no_stdin"
+    if (args.indexOf(no_stdin) == -1)
+      args.append(" ").append(no_stdin)
+
+    // need to keep reference around to prevent GC and subsequent file delete
+    if (getEnableCacheFile) {
+      val cacheFile = java.io.File.createTempFile("vowpalwabbit", ".cache")
+      cacheFile.deleteOnExit()
+
+      args.append(s" -k --cache_file=${cacheFile.getAbsolutePath} --passes ${getNumPasses}")
+    }
+
+    log.info(s"VowpalWabbit args: ${args}")
+
+    args.result
+  }
+
+  /**
+    * Internal training loop.
+    * @param df the input data frame.
+    * @param vwArgs vw command line arguments.
+    * @param contextArgs This lambda returns command line arguments that are executed in the final execution context.
+    *                    It is used to get the partition id.
+    */
+  private def trainInternal(df: DataFrame, vwArgs: String, contextArgs: => String = "") = {
+    val applyLabel = createLabelSetter(df)
+
+    val featureColIndices = generateNamespaceInfos(df.schema)
+
+    // only perform the inner-loop if we cache the inputRows
+    val numPassesInSpark = if (!getEnableCacheFile && getCacheRows) getNumPasses else 1
+
+    def trainIteration(inputRows: Iterator[Row],
```

Comment:
have you tested with empty partitions?

Reply:
in lightgbm I had to fix a bug recently by ignoring empty partitions and returning from them early on network init - it can happen often apparently with bad repartitioning from user

Reply:
can you point me to the unit test and code?